src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -138,10 +138,15 @@ public void NotifyCompletion(int hr, int bytes)
 
         public void Complete()
         {
-            lock (_contextSync)
+            lock (_context._contextLock)
             {
                 _stopped = true;
-                NativeMethods.HttpTryCancelIO(_handler);
+
+                // Should only call CancelIO if the client hasn't disconnected
+                if (!_context._clientDisconnected)
+                {
+                    NativeMethods.HttpTryCancelIO(_handler);
```

Comment:
Who invalidates this handle?

Reply:
The IISHttpContext: https://github.com/dotnet/aspnetcore/blob/master/src/Servers/IIS/IIS/src/Core/IISHttpContext.cs#L674